### PR TITLE
Batching callback

### PR DESCRIPTION
**Pull request by choshun**
This adds batch logging for batched images.

The end result is a txt file:
image

### CODE REVIEW

File: imagely.bin.js [hunk 1]
```diff
@@ -39,10 +39,16 @@ if (options.log) {
 else if (options.batch) {
 	callback = function() {
 		if (this.jsonIndex < this.json.length) {
-			var name = this.destination.split('.').slice(0, -1).pop();
-			var uniqueName = this.destination.replace(name, name + this.jsonIndex);
-			var html = this.setWindowData(this.originalHtmlString, JSON.stringify(this.json[this.jsonIndex]));
+			var json = this.json[this.jsonIndex];
+			var uniqueName;
```
Comment (choshun) [inline imagely.bin.js]: This could use some work, but will allow us to test easier.

File: imagely.bin.js [hunk 2]
```diff
@@ -39,10 +39,16 @@ if (options.log) {
...
 
+			if (json.length > 0) {
+				var name = this.destination.split('/').slice(-1).pop();
```
Comment (mpetrovich) [inline imagely.bin.js]: For filepath parsing, use Node's built-in `path` methods:
```js
var name = path.basename(this.destination);
```
You'll need to `var path = require('path');` above but don't need to `yarn add` anything—`path` is a native Node package.

File: imagely.bin.js [hunk 3]
```diff
@@ -39,10 +39,16 @@ if (options.log) {
...
+				uniqueName = this.destination.replace(name, json[0].key + '.gif');
```
Comment (mpetrovich) [inline imagely.bin.js]: `key` is a very non-obvious and non-descriptive property. Consider `filename`

File: imagely.bin.js [hunk 4]
```diff
@@ -17,6 +22,41 @@ options.bg = args.bg || args.b;
...
+	if (log.width && log.height) {
+		logs.success.push(log);
+	}
+	else {
+		logs.failure.push(log);
+	}
+}
+
+function writeLogFile(logFilepath) {
+	fs.writeFile(logFilepath, JSON.stringify(logs, 0, 4), function(err) {
```
Comment (mpetrovich) [inline imagely.bin.js]: The parameters passed to `JSON.stringify()` don't look correct: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

File: imagely.bin.js [hunk 5]
```diff
@@ -39,13 +79,19 @@ if (options.log) {
 else if (options.batch) {
 	callback = function() {
 		if (this.jsonIndex < this.json.length) {
-			var name = this.destination.split('.').slice(0, -1).pop();
-			var uniqueName = this.destination.replace(name, name + this.jsonIndex);
-			var html = this.setWindowData(this.originalHtmlString, JSON.stringify(this.json[this.jsonIndex]));
+			var json = this.json[this.jsonIndex];
+			var uniqueName;
+			var name = this.destination.split('/').slice(-1).pop();
```
Comment (mpetrovich) [inline imagely.bin.js]: This isn't a platform-compatible way to parse filepaths (eg. on Windows the delimiter isn't `/`). Use the node `path` module instead (eg. `path.basename()`).

File: imagely.bin.js [hunk 6]
```diff
@@ -39,13 +79,19 @@ if (options.log) {
...
+			var html = this.setWindowData(this.originalHtmlString, JSON.stringify(json));
+			uniqueName = this.destination.replace(name, json[0].key + '.gif');
```
Comment (mpetrovich) [inline imagely.bin.js]: Consider declaring and assigning `uniqueName` at the same time

**Review by mpetrovich — CHANGES_REQUESTED** [review]
Needs a few changes. In addition, could you add some more color to the description around what you're doing and why? Coming back to this in the future (and even now after being away from it for a few weeks), it's hard to understand what this is trying to accomplish.

File: imagely.bin.js [hunk 7]
```diff
@@ -17,6 +22,41 @@ options.bg = args.bg || args.b;
...
+	try {
+		dimensions = imageSize(destination);
+	}
+	catch (exception) {
+		dimensions = { width: null, height: null };
+	}
+
+	log = dimensions;
+	log['uuid'] = imageKeyArray[1];
+	log['endpoint'] = imageKeyArray[0];
```
Comment (mpetrovich) [inline imagely.bin.js]: Prefer object property syntax: `log.uuid = `, etc.

Comment (mpetrovich) [reply imagely.bin.js]: "endpoint" and "uuid" are _very_ NBS specific.

File: imagely.bin.js (hunk 4 above)
Comment (choshun) [reply imagely.bin.js]: Looks like for both "replacer" and "space" optional params you can use a number, this is identical to `JSON.stringify(logs, null, '\t')`.

I can change to be more readable if you want.

File: imagely.bin.js (hunk 5 above)
Comment (choshun) [reply imagely.bin.js]: Thanks, good to know. Changed.

File: imagely.bin.js (hunk 6 above)
Comment (choshun) [reply imagely.bin.js]: Yup, changed.

File: imagely.bin.js (hunk 7 above)
Comment (choshun) [reply imagely.bin.js]: I think this whole commit is very site specific (hence commit message), which brings up a broader point of maybe having this be in a separate branch, or passing in the callback altogether from the code base. I feel like this file should literally just be setting options and instantiating Imagely with options and a user set callback.

Not sure the cleanest way to do this.

**Comment by choshun** [comment]
Sure, going back to my last comment, I really think this file shouldn't need explanation, and too much site specific stuff is in here in retrospect. I can add comments, but they will be very descriptive of a problem that isn't necessarily a responsibility of Imagely.

File: imagely.bin.js (hunk 4 above)
Comment (mpetrovich) [reply imagely.bin.js]: the `0` threw me off—the replacer is supposed to be `null`, a function, or an array. Why do you need to pretty print it?

File: imagely.bin.js (hunk 3 above)
Comment (choshun) [reply imagely.bin.js]: Woops, missed this.

Sure, will change for clarity.

File: imagely.bin.js (hunk 4 above)
Comment (choshun) [reply imagely.bin.js]: I guess it's an old habit, changed to `null`.

**Review by mpetrovich — APPROVED** [review]

**Pull request merged**